src/ess/reduce/time_of_flight/toa_to_tof.py

```diff
@@ -0,0 +1,520 @@
+    sorted bin edges.
+
+    WARNING:
+    This function is highly experimental, has limitations and should be used with
+    caution. It is a workaround to the issue that rebinning data with unsorted bin
+    edges is not supported in scipp.
+    We also do not support variances on the data.
+    As such, this function is not part of the default set of providers, and needs to be
+    inserted manually into the workflow.
+
+    Parameters
+    ----------
+    da:
+        TofData with the time-of-flight coordinate.
+    """
+    events = to_events(da.rename_dims(time_of_flight="tof"), "event")
+
+    # Define a new bin width, close to the original bin width.
+    # TODO: this could be a workflow parameter
+    coord = da.coords["tof"]
+    bin_width = (coord["time_of_flight", 1:] - coord["time_of_flight", :-1]).nanmedian()
+    rehist = events.hist(tof=bin_width)
+    for key, var in da.coords.items():
+        if "time_of_flight" not in var.dims:
+            rehist.coords[key] = var
+    return ReHistogrammedTofData(rehist)
+
+
+def default_parameters() -> dict:
+    """
+    Default parameters of the time-of-flight workflow.
+    """
+    return {
+        PulseStride: 1,
+        PulseStrideOffset: 0,
+        DistanceResolution: sc.scalar(0.1, unit="m"),
+        TimeOfArrivalResolution: 500,
+        LookupTableRelativeErrorThreshold: 1.0e-2,
+        SimulationSeed: 1234,
+        NumberOfNeutrons: 1_000_000,
+    }
+
+
+def _providers() -> tuple[Callable]:
+    """
+    Base providers of the time-of-flight workflow.
+    """
+    return (
+        compute_tof_lookup_table,
+        extract_ltotal,
+        find_fastest_neutron,
+        frame_period,
+        masked_tof_lookup_table,
+        pivot_time_at_detector,
+        pulse_period_from_source,
+        time_of_arrival_folded_by_frame,
+        time_of_arrival_minus_pivot_time_modulo_period,
+        time_of_flight_data,
+        unwrapped_time_of_arrival,
+        unwrapped_time_of_arrival_minus_frame_pivot_time,
+    )
+
+
+def standard_providers() -> tuple[Callable]:
+    """
+    Standard providers of the time-of-flight workflow, using the ``tof`` library to
+    build the time-of-arrival to time-of-flight lookup table.
+    """
+    from .tof_simulation import run_tof_simulation
+
+    return (*_providers(), run_tof_simulation)
+
+
+class TofWorkflow:
+    """
+    Helper class to build a time-of-flight workflow and cache the expensive part of
+    the computation: running the simulation and building the lookup table.
+    """
+
+    def __init__(
+        self,
+        choppers,
+        facility,
+        ltotal_range,
+        pulse_stride=None,
+        pulse_stride_offset=None,
+        distance_resolution=None,
+        toa_resolution=None,
+        error_threshold=None,
+        seed=None,
+        number_of_neutrons=None,
+    ):
+        import sciline as sl
+
+        self.pipeline = sl.Pipeline(standard_providers())
```

Comment:
I suggest changing this class so the simulation providers are dependency-injected in the constructor. This will allow for easily swapping it out for a different simulation.